programmazione-funzionale.md

```diff
@@ -1 +1,198 @@
 # Programmazione funzionale
+
+Il concetto di programmazione funzionale si basa sul fatto di considerare le funzioni come valori, e quindi utilizzarle come tali. In questo modo, le funzioni possono essere passate come parametri, ritornate come risultati, memorizzate in variabili, ecc.
+Le funzioni usate come parametro senza specificare il nome della funzione, ma solo il comportamento che essa deve avere, sono dette **funzioni anonime**.
+
+## Espressioni lambda
+
+Supponiamo di avere una classe fatta in questo modo:
+
+```java
+class Persona {
+    private String nome;
+    private int eta;
+
+    public Persona(String nome, int eta) {
+        this.nome = nome;
+        this.eta = eta;
+    }
+
+    public String getNome() {
+        return nome;
+    }
+
+    public int getEta() {
+        return eta;
+    }
+}
+```
+
+e supponiamo di avere una lista di persone: 
+
+```java
+List<Persona> persone = new ArrayList<>();
+persone.add(new Persona("Mario", 20));
+persone.add(new Persona("Luigi", 30));
+persone.add(new Persona("Pippo", 40));
+persone.add(new Persona("Pluto", 50));
+```
+
+Possiamo, ad esempio, ordinare la lista di persone per età, usando il metodo `sort` della classe `Collections`: 
+```java
+public static <T> void sort(
+    List<T> list, Comparator<? super T>
+) c
+```
+
+Il metodo `sort` prende come primo parametro una lista di oggetti di tipo `T`, e come secondo parametro un oggetto di tipo `Comparator`, possiamo passare un `Comparator` direttamente usando una **espressione lambda**:
+
+```java
+obj.method(
+    (param1, param2) -> {
+        // corpo della funzione
+    }
+);
+```
+
+Nel nostro esempio il metodo `sort` sarà implementato così:
+
+```java
+Collections.sort(persone, (p1, p2) -> {
+    /* Java conosce già il tipo di p1 e p2, 
+    quindi non è necessario specificarlo */
+    if(p1.getEta() > p2.getEta()) {
+        return 1; }
+    else if(p1.getEta() < p2.getEta()) {
+        return -1; }
+    else return 0;
+});
+```
+
+## Composizione di funzioni (`Stream<T>`)
+
+Possiamo concatenare delle funzioni che agiscono su una Collezione per costruire una **catena di funzioni**, che filtrerà, mapperà e agirà sulla nostra collezione.
+
+Questa catena inizierà invocando il metodo **`stream()`** sulla collezione. Nell'esempio:
+
+```java
+Stream<Persona> stream = persone.stream();
+```
+
+### `Filter`
+
+Possiamo filtrare lo `stream` per far si che tutti gli elementi della collezione soddisfino un cetro predicato (funzione che ritorna solo true o false).
```

Comment:
Se spieghi cos'è il predicato prima, credo si possa rimuovare la parte tra parentesi

Reply:
Non l'ho spiegato prima, credo le l'elenco Predicato, Funzione, ... (quello che c'è dalle slide) sia scontato o per lo meno scomodo all'esame (in cui devi avere tutto sottomano subito). Quindi l'ho spiegato "on the fly"